src/DocumentDB.ChangeFeedProcessor/FeedProcessing/ChangeFeedQueryTimeoutDecorator.cs

```diff
@@ -0,0 +1,49 @@
+﻿// ----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  Licensed under the MIT license.
+// ----------------------------------------------------------------
+
+namespace Microsoft.Azure.Documents.ChangeFeedProcessor.FeedProcessing
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.DataAccess;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.Monitoring;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.Utils;
+    using Microsoft.Azure.Documents.Client;
+
+    internal class ChangeFeedQueryTimeoutDecorator : IChangeFeedDocumentQuery<Document>
+    {
+        private readonly IChangeFeedDocumentQuery<Document> query;
+        private readonly IHealthMonitor monitor;
+        private readonly TimeSpan timeout;
+
+        public ChangeFeedQueryTimeoutDecorator(IChangeFeedDocumentQuery<Document> query, IHealthMonitor monitor, TimeSpan timeout)
+        {
+            this.query = query;
+            this.monitor = monitor;
+            this.timeout = timeout;
+        }
+
+        public bool HasMoreResults => this.query.HasMoreResults;
+
+        public async Task<IFeedResponse<TResult>> ExecuteNextAsync<TResult>(CancellationToken token = default(CancellationToken))
```

Comment:
Why do we need the decorators and wrappers? Can't we just add the AbortOnTimeout extension on the PartitionProcessor? Or this is because the PartitionProcessor can have a custom implementation?

Reply:
yes, I was thinking that PartitionProcessor can have custom implementation + discussed with @mkolt and he wanted to have this logic outside of processor.
I personally think a wrapper can be handy if we want to report more thing related to DocumentClient calls to health monitor. If not, then it's probably not very useful